### ISSUE TITLE
SpriteBatch With Custom Effects and Per-Sprite Data

### ISSUE THREAD

**Issue by Tommy-Hu** [MonoGame/MonoGame#8295]
### Intent

Allow per-sprite data (that comes along with each `Draw()` call) to be easily passed into effects -- without beginning new batches or needing to use `DrawUserIndexedPrimitive()`. The user will be able to access the per-sprite data in the Vertex Shader of the effect, and can therefore easily perform per-sprite effects. Having extra data per vertex will increase batch item size, but the user will be able to choose whether or not they need to use this new functionality (by using an old overload or setting the new overload's parameter to its default value).

## Possible Solution
Create an overload `SpriteBatch.Draw(OTHER_OLD_PARAMS, params float[] customFloats)`. The framework throws an error if `customFloats.Length` exceeds some threshold (so that the user cannot put in a `float[]` of length 100000 for fun) or if the given `customFloats` array is `null`.

A possible implementation of the new `SpriteBatch.Draw` overload might be one of the following (I only dug around the MonoGame code a little bit, so these may be impossible):

1. There would be a few new MonoGame internal `struct`s created, each representing a custom vertex of a set number of floats (e.g., a `VertexPositionColorTextureF1`, `VertexPositionColorTextureF2`, `VertexPositionColorTextureF3`,... that supports 1, 2, and 3 extra floating-point data on vertices respectively). The new structs would extend `IVertexType`. The new `Draw` overload adds one of these into the batch depending on the length of `customFloats`.

2. A new MonoGame internal `struct VertexPositionColorTextureSingles : IVertexType` can be created that dynamically changes its `VertexDeclaration` size based on the number of extra floating parameters to the `Draw` overload.

3. If the SpriteBatch must send batches of vertices of the same struct size, then the batcher could either send the batch of vertices based on the maximum Length of `customFloats` stored in the batch or instead, add a new optional parameter to `SpriteBatch.Begin()` that specifies the number of extra floats to carry per vertex/`Draw` call.

Either way, this solution would achieve the goal of easily sending custom data to each effect per sprite, essentially allowing different effect parameters in a single batch.

### Motivation

2D games that need to use multiple effects may find beginning a new sprite batch for each effect very slow (this is what happened in my game). A way to achieve multiple effects with fewer sprite batches is to write a master effect that is a collection of all effects in one single effect file. A custom vertex (that extends `IVertexType`) can then be used to send an extra per-vertex floating-point identifier variable to the vertex shader, which can then choose to apply one or more of the sub-effects of the master effect based on the received identifier. Right now, there is no way that I know of that allows custom vertices in SpriteBatch. To achieve the effect indexing described above, `GraphicsDevice.DrawUserIndexedPrimitives()` would be used, meaning that the user would need to remake the SpriteBatch's sorting functionality, matrix calculations, etc.

The user may also wish to send other per-sprite data to the vertex shader. Beginning a new `SpriteBatch` to set uniform shader variables **for each sprite that has a different shader data** or manually drawing the object vertices with `DrawUserIndexedPrimitives()` are both unattractive options, as mentioned above.

**Comment by stromkos** [comment]
Set a `uniform` variable in the shader, modify the value between draw calls.

---

> 2D games that need to use multiple effects may find beginning a new sprite batch for each effect very slow

Not if they are batched properly.  Group all Draw calls by Effect.

**Comment by Tommy-Hu** [comment]
Hey @stromkos, thanks for the suggestion :) However, the problem with `uniform` and grouping `Draw` calls by effect is that doing so will disrupt the ordering of the sprites.

For example, in a top-down game, the dev may want to have sprites with different effects for objects, but those objects must be grouped by the Y component instead of by effect. In particular, let's say the dev wants all the enemies to use a particular shader, and every other object including the player, vegetation, buildings etc. uses some other shader. Now, since the game is top-down, the sort order must be by the Y component so that objects behind other objects are properly rendered.

---

Here is a more concrete example. Please bear with my terrible handwriting.

![image](https://github.com/MonoGame/MonoGame/assets/64673136/e17b6b93-2b34-45d4-a2fa-bf1ba7635edb)

If the player or animals walk behind a building, the building will be drawn after the player and/or the animals so they don't appear above the building. Of course, the opposite is true if the player or animals walk in front of the building, and the building sprite must be rendered before the animals and/or the player sprites. In this case, if the dev wants to add a glow effect for the player and the animals, but no effect for the building, then the player cannot be grouped with the glowing animals. Specifically, if the player is behind the building but the animals are in front of the building, grouping sprites by glow effect causes the animals and the player to be rendered together, either before or after the building sprite renders. The player will then see either nothing because the building renders after the glowing sprites, or see the player on top of the building because the building renders before the glowing sprites.

Of course, as I suggested before, a way to solve this is to pass per-sprite data into the vertex shader of a master effect to select whether a sprite uses the glow effect or not at runtime.

### TLDR
The only way I know to render a top-down game for an arbitrary number of sprites is to sort them by their Y. This conflicts with sorting sprites by `Effect`. Now, even if all sprites use the same effect, every sprite may have different effect parameters. For instance, different glowing sprites have different `glowAmount` and `glowColor` values. This cannot be done with `uniform` variables unless I begin a new batch for each sprite.

**Comment by stromkos** [comment]
Group all Draw calls by Effect and pass custom `Depth` value to the renderer per object.

**Comment by tomspilman** [comment]
I think some option to pass a fixed per-sprite data to the shader isn't a bad idea.

But for it to be per-sprite and efficient it may need to be part of the vertex format or use instancing to pass per-instance data to the sprite.  Instancing seems to make the most sense with a seperate VB with per-instance data managed by `SpriteBatch`.

So it could go like:

```
batch.Begin();

// Set the instance data for the next draws.
batch.SetInstanceData(x);

// Draw two sprites with the same per-instance data.
batch.Draw(mySprite);
batch.Draw(mySprite);

batch.End();
```

Thoughts?

**Comment by Tommy-Hu** [comment]
Hey @tomspilman, thanks for the reply!

I like the idea of using instancing, but I'm trying to understand how `SetInstanceData()` would work with arbitrary sprite ordering. My main concern is whether calling `SetInstanceData()` with different values would force a batch flush.

The core problem I'm trying to solve:
In a Y-sorted top-down game, I need to draw sprites in strict Y-coordinate order (regardless of their shader parameters) to get proper depth layering. For example:
```csharp
batch.Begin(SpriteSortMode.BackToFront);
batch.Draw(player, y: 150, glowAmount: 1.0f);    // needs lots of glow effect
batch.Draw(building, y: 200, glowAmount: 0.0f);  // no glow effect
batch.Draw(animal, y: 250, glowAmount: 0.8f);    // needs some glow effect
batch.End();
```

These three sprites need different shader parameters but **must be drawn in Y order** within a single batch.

My question about `SetInstanceData()`:
Would this approach require flushing the batch every time `SetInstanceData()` is called with new data? If so, wouldn't that break the batch for each unique parameter combination, essentially forcing us back to either:
* Multiple Begin/End pairs grouped by effect parameters (losing Y-sort), or
* One Begin/End pair per sprite with different instance data (losing batch efficiency)?

Or are you suggesting that `SpriteBatch` would internally buffer multiple instance data sets and submit them all in one draw call? That would be awesome! But I'm not sure how that would work with the current `SpriteBatch` architecture.

Why vertex data seemed appealing:
Passing data directly through the vertex format (like `VertexPositionColorTextureCustom`) would let `SpriteBatch` handle everything the same way it already does - just with larger vertices. The sorting, batching, and flushing logic wouldn't need to change at all.

Let me know if I'm misunderstanding something, and I would love to hear your thoughts on how the instancing approach would handle this use case!

**Comment by mghoffmann** [comment]
I agree there would be a lot of value in passing extra information to `SpriteBatch.Draw` to allow custom shaders to change behaviors for different sprites.

I don't think the right way to do that is to pass data for each vertex, because right now `SpriteBatch` is very tightly coupled to `VertexPositionColorTexture`. It's internal `SpriteBatcher` helper makes a quad for each sprite passed to `Draw`, with a different `VertexPositionColorTexture` for each corner of the quad. So the vertex buffer used in the call to `GraphicsDevice.DrawUserIndexedPrimitives` has `SpriteCount * 4` vertices, forming `SpriteCount * 2` triangle primitives in a `TriangleList`.

This is different from one method of batched/instanced rendering where the vertex buffer would only contain 4 vertices (a quad centered around `(0,0,0)`), but a second vertex buffer would contain `SpriteCount` "vertices" that are actually data about each instance, like translation, rotation, scale, color, texture clip, etc.

Moving `SpriteBatch` to use that method of instanced rendering would require huge changes as well, but it might be worth it if we could define our own instance data structs and pass them in with `Draw`. Doing nearly the same changes to support `VertexPositionColorTextureF1`, `F2`, etc. would be similar but would bind users to selecting 1 to N extra floats for their use case, which might not always make sense for the shader parameters (which might be int, bool, etc. instead of float). I think a generic overload of Draw that takes an instance data struct as a type parameter would be more flexible.

```csharp
struct MyInstanceData : IVertexType
{
    // World coordinates
    public Vector3 Position;
    // Texture coordinates
    public Vector2 TopLeft;
    public Vector2 BottomRight;
    public float Depth;
    public float GlowAmount;

    public VertexDeclaration VertexDeclaration { get; } = new VertexDeclaration(
        new VertexElement(0, VertexElementFormat.Vector3, VertexElementUsage.Position, 0),
        new VertexElement(sizeof(float) * 3, VertexElementFormat.Vector2, VertexElementUsage.TextureCoordinate, 0),
        new VertexElement(sizeof(float) * (3 + 2), VertexElementFormat.Vector2, VertexElementUsage.TextureCoordinate, 1),
        new VertexElement(sizeof(float) * (3 + 2 + 2), VertexElementFormat.Single, VertexElementUsage.Depth, 0),
        new VertexElement(sizeof(float) * (3 + 2 + 2 + 1), VertexElementFormat.Single, VertexElementUsage.Depth, 1),
    );
}


// Then to draw an instance:
spriteBatch.Draw<MyInstanceData>(playerTexture, playerInstanceData);
```

Things would still need to be sub-batched by instance data type, but `SpriteBatcher` could be building those batches between `Begin` and `End`/flush.

**Comment by Tommy-Hu** [comment]
Hi @mghoffmann let me know if I understand correctly:
1. 1 shared quad in vertex buffer (4 vertices total, reused for all sprites)
1. 1 instance data entry per sprite in a separate instance buffer
1. Each instance contains position, texture coords, color, depth, AND any user-defined custom data (like GlowAmount).
1. User defines what the custom data is by implementing a custom `IVertexData`.
1. User passes custom data using a generic `Draw` overload that only takes a texture and an instance data (since the instance data should already contain info about position, uv etc)
1. `SpriteBatch` internally creates sub-batches when `instanceData.GetType()` changes.
1. This only works if user writes custom vertex shader that matches the fields in instanceData along with the vertex information from the shared quad. Example:
  ```hlsl
struct VSInput {
    // From the shared quad vertex buffer (per-vertex data)
    float3 QuadPosition : POSITION0;
    float2 QuadTexCoord : TEXCOORD0;
    
    // From the instance buffer (per-instance data)
    float3 InstancePosition : POSITION1;
    float2 InstanceScale : TEXCOORD1;
    float Rotation : TEXCOORD2;
    float GlowAmount : TEXCOORD3;
};
  ```

I think this is a great idea!

**Comment by mghoffmann** [comment]
Yes, that's a good summary of what I'm thinking. This is how I'm doing instanced rendering with 3D meshes right now, and sprite quads are just really small 3D meshes 😆

I think this would be a big breaking change to the XNA API, or at least significantly change how SpriteBatch works internally, so it's probably something to save for a 4.0.0+ release.

This generic batching could also be extended to 3D drawing, by letting the library user specify the mesh `VertexBuffer` and `IndexBuffer` to define shapes other than flat quads. If I were to design this from scratch I would have `MeshBatch` as a base class that takes a `VertexBuffer`, an `IndexBuffer`, and an `InstanceData` for a `Draw` method. Then `SpriteBatch` would become a derived type of `MeshBatch` that hid the `Draw` method with its own quad vertex and index buffer.

```csharp
public interface IVertexProvider : IDisposable
{
    public Type VertexType { get; }
    public VertexBuffer VertexBuffer { get; }
    public IndexBuffer? IndexBuffer { get; }
}

// In SpriteBatch.cs

private IVertexProvider quadVertexProvider = new QuadVertexProvider<VertexPositionColorTexture>();
new public void Draw<TInstanceData>(Texture2D texture, TInstanceData data) => base.Draw<TInstanceData>(quadVertexProvider, data, texture);
```

Right now `SpriteBatcher`'s drawing also sets the texture directly to slot 0 on the graphics device, which is very hand-holdy and could be changed to require the caller to set up device resources such as textures, or provide references to the resources with instance data as the key for another round of sub-batching.

https://github.com/MonoGame/MonoGame/blob/3f4423b7414a7f3347959e843d7c7c29281dde2f/MonoGame.Framework/Graphics/SpriteBatcher.cs#L257-L259

But then again the whole point of `SpriteBatch` is to hold your hand for getting sprites on the screen without a lot of boilerplate. So maybe this whole instanced rendering idea is better for an extension library project, or instead of `SpriteBatch` a new class that relies on the user to set up the Effect and graphics resources for their bespoke shaders before beginning the batch.

**Comment by Tommy-Hu** [comment]
Woah, cool stuff! I agree with how adding the `SpriteBatch.Draw<T>` overload would introduce a breaking change in internal code, and I think maybe having an extension would work better to start as a prototype? Do you think it's worth making an extension library first to prototype/explore this? Maybe we can make a super simple (and potentially buggy and unoptimized) `InstancedSpriteBatch` or similar class, and if it works well in the extension library, then we can consider adding either the `InstancedSpriteBatch` class to MonoGame directly or modify the existing `SpriteBatch` class.

Regardless, I agree that `SpriteBatch` already reduces a lot of complexity for MonoGame users (since even a beginner can quickly draw some 2D stuff on screen in a couple minutes). However, in my opinion, MonoGame is not a _very_ "hand-holdy" library for making games (since game engines definitely reduce a lot more complexity), which is why I think MonoGame users could benefit from having more control over the rendering system. I mean, part of the reason I chose MonoGame a few years ago over something like Unity or Godot is to build my own rendering system that is tailored/optimized to my games. (By the way, I have not used XNA before lol, so I'm not sure how different MonoGame is from XNA). But then again, I guess you can always use primitive functions, but I feel like it's all or nothing right now -- either use SpriteBatch which is very rigid but convenient, or use primitive functions that requires a LOT of setup to manually batch.

Let me know your thoughts :) !
